graphql_compiler/query_pagination/query_parameterizer.py

```diff
@@ -1,61 +1,288 @@
+        return True
+    if filter_operation_1 == ">=" and filter_operation_2 == ">=":
+        return True
+    return False
+
+
+def _add_pagination_filter_at_node(
+    query_ast: DocumentNode,
+    pagination_field: str,
+    directive_to_add: DirectiveNode,
+    extended_parameters: Dict[str, Any],
+) -> Tuple[DocumentNode, Dict[str, Any]]:
+    """Add the filter to the target field, returning a query and its new parameters.
+
+    Args:
+        query_ast: The query in which we are adding a filter
+        pagination_field: The field on which we are adding a filter
+        directive_to_add: The filter directive to add
+        extended_parameters: The original parameters of the query along with
+                             the parameter used in directive_to_add
+
+    Returns:
+        tuple (new_ast, removed_parameters)
+        new_ast: A query with the filter inserted, and any filters on the same location with
+                 the same operation removed.
+        new_parameters: The parameters to use with the new_ast
+    """
+    if not isinstance(query_ast, (FieldNode, InlineFragmentNode, OperationDefinitionNode)):
+        raise AssertionError(
+            f'Input AST is of type "{type(query_ast).__name__}", which should not be a selection.'
+        )
+
+    new_directive_operation = _get_filter_node_operation(directive_to_add)
+    new_directive_parameter_name = _get_binary_filter_node_parameter(directive_to_add)
+    new_directive_parameter_value = extended_parameters[new_directive_parameter_name]
+
+    # If the field exists, add the new filter and remove redundant filters.
+    new_parameters = dict(extended_parameters)
+    new_selections = []
+    found_field = False
+    for selection_ast in query_ast.selection_set.selections:
+        new_selection_ast = selection_ast
+        field_name = get_ast_field_name(selection_ast)
+        if field_name == pagination_field:
+            found_field = True
+            new_selection_ast = copy(selection_ast)
+            new_selection_ast.directives = copy(selection_ast.directives)
+
+            new_directives = []
+            for directive in selection_ast.directives:
+                operation = _get_filter_node_operation(directive)
+                if _is_filter_redundant(new_directive_operation, operation):
+                    parameter_name = _get_binary_filter_node_parameter(directive)
+                    parameter_value = new_parameters[parameter_name]
+                    if not _is_new_filter_stronger(
+                        operation, new_directive_parameter_value, parameter_value
+                    ):
+                        raise AssertionError(
+                            f"Pagination filter {directive_to_add} on "
+                            f"{pagination_field} is not stronger than "
+                            f"an existing filter {directive}. This is "
+                            f"likely a bug in parameter generation."
```

Comment:
Maybe print the query string in addition to all this, so we can reproduce more easily?